core-bundle/src/Command/MigrateCommand.php

```diff
@@ -0,0 +1,278 @@
+    {
+        $this->migrations = $migrations;
+        $this->fileLocator = $fileLocator;
+        $this->projectDir = $projectDir;
+        $this->framework = $framework;
+        $this->installer = $installer;
+
+        parent::__construct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure(): void
+    {
+        $this
+            ->setName('contao:migrate')
+            ->addOption('complete', null, InputOption::VALUE_NONE, 'Execute all database migrations including DROP queries. Can be used together with --no-interaction.')
+            ->addOption('schema-only', null, InputOption::VALUE_NONE, 'Execute database schema migration only.')
+            ->setDescription('Executes migrations and the database schema diff.')
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->io = new SymfonyStyle($input, $output);
+
+        if ($input->getOption('schema-only')) {
+            return $this->executeSchemaDiff($input->getOption('complete')) ? 0 : 1;
+        }
+
+        if (!$this->executeMigrations()) {
+            return 1;
+        }
+
+        if (!$this->executeSchemaDiff($input->getOption('complete'))) {
+            return 1;
+        }
+
+        if (!$this->executeMigrations()) {
+            return 1;
+        }
+
+        $this->io->success('All migrations completed.');
+
+        return 0;
+    }
+
+    private function executeMigrations(): bool
+    {
+        while (true) {
+            $first = true;
+
+            foreach ($this->migrations->getPendingMigrations() as $migration) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * '.$migration);
+            }
+
+            $runOnceFiles = $this->getRunOnceFiles();
+
+            if ($runOnceFiles) {
+                @trigger_error('Using runonce files has been deprecated and will no longer work in Contao 5.0. Use the migration framework instead.', E_USER_DEPRECATED);
+            }
+
+            foreach ($runOnceFiles as $file) {
+                if ($first) {
+                    $this->io->section('Pending migrations');
+                    $first = false;
+                }
+
+                $this->io->writeln(' * Runonce file: '.$file);
+            }
+
+            if ($first) {
+                break;
+            }
+
+            if (!$this->io->confirm('Execute the listed migrations?')) {
+                return false;
+            }
+
+            $this->io->section('Execute migrations');
+
+            foreach ($this->migrations->runMigrations() as $result) {
+                $this->io->writeln(' * '.$result->getMessage());
+
+                if (!$result->isSuccessful()) {
+                    $this->io->error('Migration failed');
```

Comment:
if a migration fails, can I tell which one it was? And do we continue with all other migrations (given they have priority)?

Reply:
> can I tell which one it was?

Yes, the message of the result gets written to the output three lines above.

> And do we continue with all other migrations

Yes, we do, as long as the migration fails “nicely” by returning an unseccessful result. If the migration wants to cancel the whole process it should throw an exception.